### PR TITLE
Stabilize test ProcessServiceAT

### PR DESCRIPTION
**Pull request by hung-nguyen-hoang**
Stabilize test ProcessServiceAT:
- On RAT instances CPU is quite busy which sometimes causes gcf task creation timeout.
- Workaround by retrying GIT deployment process 3 times, only fail tests after 3 attempts.

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-java/src/test/java/com/gooddata/sdk/service/RetryUtil.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    public static <T> T retry(Supplier<T> function) {
+        int retryCount = 0;
+        T result = null;
+        boolean success = false;
+        while (!success && retryCount < MAX_RETRIES) {
+            try {
+                result = function.get();
+                success = true;
+            } catch (Exception ex) {
```
Comment (danh-ung) [inline gooddata-java/src/test/java/com/gooddata/sdk/service/RetryUtil.java:24]: we should sleep 5 seconds before retry.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-java/src/test/java/com/gooddata/sdk/service/RetryUtil.java (hunk 1 above)
Comment (hung-nguyen-hoang) [reply gooddata-java/src/test/java/com/gooddata/sdk/service/RetryUtil.java:24]: Done fix it

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by liry** [comment]
Plz, remove "BUGFIX: MSF-18209" prefix from commit msg - it's useless in this public repo. And add explanation notes same as in PR also as next lines of commit msg.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by hung-nguyen-hoang** [comment]
Hi @liry Done remove prefix and add explanation notes for commit message. Thanks you.

**Comment by liry** [comment]
Do we really want to retry these POST/PUT requests (consider idempotence and safety of these HTTP methods)? All response codes? If some (sub)set is safe and idempotent, shouldn't we implement it into service instead of just into tests?

**Comment by danh-ung** [comment]
@liry these POST requests are idempotence and safety. The issue only happens in PI, no need to invest retries in the service.

File: gooddata-java/src/test/java/com/gooddata/sdk/service/dataload/processes/ProcessServiceIT.java [hunk 2]
```diff
@@ -330,8 +331,9 @@ public void shouldCreateProcessFromAppstorePackageRegistryMiss() throws Exceptio
              .respond()
                 .withStatus(200)
                 .withBody(readFromResource("/dataload/processes/appstoreProcess.json"));
-        DataloadProcess process = gd.getProcessService().createProcessFromAppstore(project, new DataloadProcess("appstoreProcess", ProcessType.RUBY.toString(),
-                "${PUBLIC_APPSTORE}:tag/prodigy-testing:/test/rubyHello")).get();
+        DataloadProcess newProcess = new DataloadProcess("appstoreProcess", ProcessType.RUBY.toString(),
+                "${PUBLIC_APPSTORE}:tag/prodigy-testing:/test/rubyHello");
+        DataloadProcess process = retry(() -> gd.getProcessService().createProcessFromAppstore(project, newProcess).get());
```
Comment (peter-plochan) [inline gooddata-java/src/test/java/com/gooddata/sdk/service/dataload/processes/ProcessServiceIT.java:336]: These tests are running against mocked backends (using Jadler), so no retry is needed in this test class.

**Comment by peter-plochan** [comment]
@hung-nguyen-hoang I'm not sure if this solution is the right one.
First, the acceptance test won't be 1:1 with the service. Imagine you have a bug in processes API which occurs only on the first try (and it won't be even discovered by our internal GoodData tests). The acceptance tests will pass, but the service itself will fail with the same backend configuration.
Second, I consider this as some workaround which should be solved in a better way - e.g. raising GCF timeout on GoodData RAT instances when running these tests or even not running Java SDK acceptance tests against RAT instances.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by danh-ung** [comment]
@pagi thank you for your comments. To your concerns:
- I couldn't imagine which case the process API will fail only on the first try, I would take that risk (if any).
- raising GCF timeout or not running Java SDK acceptance tests against RAT instances would take much more effort than this. I don't think we would like to invest to introduce timeout param in GCF component or changing pipeline to not use RAT instances.

File: gooddata-java/src/test/java/com/gooddata/sdk/service/dataload/processes/ProcessServiceIT.java (hunk 2 above)
Comment (hung-nguyen-hoang) [reply gooddata-java/src/test/java/com/gooddata/sdk/service/dataload/processes/ProcessServiceIT.java]: @peter-plochan Done update code. Many Thanks.

**Comment by liry** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-java-merge-compliance-check/175/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-java-fossa-scanning-zuul-el7/15/ : SUCCESS in 14s (non-voting)

**Pull request merged**